### ISSUE TITLE
Expanding the Syntax Extractor

### ISSUE THREAD

**Issue by daomcgill** [sailuh/kaiaulu#313]
-----

## Purpose 

The Syntax Extractor in Kaiaulu is used to extract meaningful information from source code using srcML. The purpose of this task is to extend the syntax extraction capabilities by adding new functions to extract file-level and class-level documentation using XPath queries. The extracted data will be used in future stages, where the goal is to combine the data with NLP to create semantic representations of the code.

## Process
1. Understand the current syntax extractor functions, and how annotations are represented in srcML XML files.
2. Understand XPath queries, and create queries for the new functions (file and class level documentation).
3. Implement new functions with the custom XPath queries.
4. Test on various examples.
5. Consider other syntactic elements that could be useful for the bigger picture. Example: Functions in git.R that retrieve commit messages or issue discussions tied to a file.
6. Create a notebook for Syntax Extraction, and maintain it with existing/ new functionality.

## Existing Functions
- [annotate_src_text()](https://github.com/sailuh/kaiaulu/blob/c7811069ef3c7943b509c76d4b9116d1a14dc343/R/src.R#L296): Runs srcML on a folder of source code and outputs the annotated XML.
- [query_src_text()](https://github.com/sailuh/kaiaulu/blob/c7811069ef3c7943b509c76d4b9116d1a14dc343/R/src.R#L327): Runs an XPath query on the annotated XML generated from annotate_src_text().
- [query_src_text_class_names()](https://github.com/sailuh/kaiaulu/blob/c7811069ef3c7943b509c76d4b9116d1a14dc343/R/src.R#L356): Extracts class names from the annotated source code.
- [query_src_text_namespace()](https://github.com/sailuh/kaiaulu/blob/c7811069ef3c7943b509c76d4b9116d1a14dc343/R/src.R#L399): Extracts the namespace (file path) from the annotated source code.

## New Functions
- query_src_text_file_documentation(): Extract file-level documentation (e.g. comments in the file header).
- query_src_text_class_documentation(): Extract class-level documentation (e.g. comments before class declaration).

## Task List 
- [ ] Install and set up [srcML](https://www.srcml.org/doc/java_srcML.html). This takes source code and add XML annotations (e.g. classes, methods, variables, etc.). Include the path to srcML in tools.yml.
- [ ] Use the existing functions to generate XML files. Inspect the files to understand the output.
- [ ] Gain an understanding of how XPath queries are formed, and figure out how to create queries for the new functions.
- [ ] Write the new XPath queries.
- [ ] Write the new functions.
- [ ] Verify that the functions work correctly.
- [ ] Consider what other functions could be useful, and implement those.
- [ ] Maintain a notebook for Syntax Extraction.

### References
- [Syntax Extractor in Kaiaulu](https://github.com/sailuh/kaiaulu/blob/c7811069ef3c7943b509c76d4b9116d1a14dc343/R/src.R#L280-L576)
- [API for Syntax Extraction](http://itm0.shidler.hawaii.edu/kaiaulu/reference/index.html#-syntax-extraction-)
- [Example XPath Query](https://github.com/sailuh/kaiaulu/blob/c7811069ef3c7943b509c76d4b9116d1a14dc343/R/src.R#L360)
- [Issue #206 Add Source Code Text Parser Module](https://github.com/sailuh/kaiaulu/issues/206)
- [Text GoF Patterns Notebook (see what srcML does for this](http://itm0.shidler.hawaii.edu/kaiaulu/articles/text_gof_showcase.html)
- [srcML slides](https://www.srcml.org/tutorials/overview.html)

-----

**Comment by daomcgill** [comment]
@carlosparadis here is my first pass at the issue for part I.

**Comment by carlosparadis** [comment]
Yes, this is great! 

The one thing I would emphasize is:

> Understand XPath queries, and create queries for the new functions (file and class level documentation).

The better mindset is to consider:

1.  Imagine a table that every row is a file in the project. The first column is the filepath, the second column contain in the cell all of its content as words. 
2. If we wrote programs purely in english, then every word would be meaningful. But the words "for", "if", etc, do not carry a lot of meaning on its own. So, in another column, we could consider creating a cell without them (let's skip this for now). 
3. Then in a subsequent column we could say, "let's just focus on docstrings since that's actual english". Now you have another column that only contain words coming from docstrings.
4. You could also have another column that in addition to the docstrings, include the variable names, since they carry meaning. Or function names, etc.

Every column is a different "extract" from all the words in each file. The functions you will add, will extract these words from a project files. This would enable the rest of the pipeline to experiment with which set of words is more meaningful.

**Comment by daomcgill** [comment]
@carlosparadis I added a notebook for syntax extraction. Could you skim it and see if this is what you are looking for? The section on XPath queries is definitely a WIP, it was more to gather my own understanding. It will be updated as I work on getting the actual queries needed for the new functions. New config file is a placeholder as of now.

**Comment by carlosparadis** [comment]
@daomcgill did you open a PR for it?

**Comment by daomcgill** [comment]
@carlosparadis I did something wrong with the PR. I just opened one [here](https://github.com/sailuh/kaiaulu/pull/320).

**Comment by daomcgill** [comment]
@carlosparadis I made the changes we discussed. I also added functions for extracting function names (opt. params) and variable names (opt. types). I do need to do a more thorough manual check to make sure the results are accurate, but it appears to be working so far.
Should I continue adding functions--maybe imports or some way of determining inheritance structure (I would need to think about this some more)--or move on and come back once the other pieces start to come together?

**Comment by carlosparadis** [comment]
@daomcgill the code that @RavenMarQ did already determines dependency types, so no need to go after that. If you already covered variables, functions, classes, is there anything else on the source code you thing carries semantic meaning? 

Also, you did this for Java, right? What were the other available languages srcML covered again?  

And lastly, did you do the parser functions for these too or just the command to create the XML?

**Comment by daomcgill** [comment]
@carlosparadis I do not think operator, specifier or control statements would be particularly meaningful. Do you have any ideas of what might be useful? The only thought I had was imports.

I did this for Java. It looks like srcML is also available for C, C++ and C#.

I am not sure what you mean by parser functions vs command to create the XML. I created functions that query the XML file generated by the preexisting function, and extract the class/file level comments, functions and variables.

**Comment by carlosparadis** [comment]
Hi Dao, 

I honestly can't think of anything else. I would say go ahead and add the imports. We can just stick with Java for now. Python would have been nice if they covered.

Once this issue is done, I suggest we also skip ahead to play with the Python Notebook that process this data on the other repo, since this will give you a better idea how this data is being used, and then we can cycle back to the issue that represent files as commit and comments associated to it. 

Have you been using your functions to parse a table of 1 file, or the entire project out? We will want to ideally get a table that contains the filepaths the classes, methods, docstring belongs so it helps with subsetting when we move to the Python script. 

I recall in the Text GoF notebook I had to create a filepath and a classpath to be able to connect both together, but maybe the xpath can give you both.

**Comment by daomcgill** [comment]
@carlosparadis I will add the imports. It looks like srcML should be adding support for more languages soon, but we do not know when.

So far, I have been using the functions on one specific src folder in the Maven repo. It contains multiple files, but not the entire project. I can work on that. The tables generated show filepath + the element that specific function is querying, e.g. filepath and classes. I could make a function that calls each of the query functions and compiles that extracted data in a single table.
